File: cpp/ems-it/src/slice_main.cpp
// ems-slice --input <journal> --output <journal> [--seed <n>]
//
// A pure function from an input journal to an output journal. No network, no
// clock, no filesystem beyond those two paths. The Java and Rust binaries
// accept the same arguments and must produce byte-identical output — see
// conformance/README.md.
//
// Exit codes: 0 success, 1 the input journal could not be read, 2 a usage
// error. These are part of the contract: the harness uses them to tell "this
// implementation rejected the input" apart from "this implementation was
// invoked wrongly".

#include <charconv>
#include <cstdint>
#include <filesystem>
#include <iostream>
#include <optional>
#include <span>
#include <string>
#include <string_view>
#include <system_error>
#include <vector>

#include "ems_core/ids.hpp"
#include "ems_core/journal.hpp"
#include "ems_it/slice_runner.hpp"

namespace {

constexpr int kExitOk = 0;
constexpr int kExitInputError = 1;
constexpr int kExitUsage = 2;

constexpr std::string_view kUsage =
    "usage: ems-slice --input <journal> --output <journal> [--seed <n>]\n"
    "\n"
    "  --input   input event journal (JSONL)\n"
    "  --output  output event journal (JSONL), overwritten\n"
    "  --seed    identifier generator seed, default 0\n"
    "\n"
    "A pure function from input journal to output journal. See conformance/README.md.\n";

struct Args {
    std::filesystem::path input;
    std::filesystem::path output;
    std::uint64_t seed{0};
};

int usage(std::string_view message) {
    if (!message.empty()) {
        std::cerr << "ems-slice: " << message << '\n';
    }
    std::cerr << kUsage;
    return kExitUsage;
}

/// STUDY: span-at-boundaries
///
/// std::span<const char* const> rather than (int argc, char** argv): the
/// boundary carries its own length, so no loop in this function can walk off
/// the end of the array by reading one argument too many.
std::optional<Args> parse_args(std::span<const char* const> argv, std::string& error) {
    Args args;
    bool have_input = false;
    bool have_output = false;

    for (std::size_t i = 0; i < argv.size(); ++i) {
        const std::string_view arg(argv[i]);
        auto next = [&](std::string_view what) -> std::optional<std::string_view> {
            if (i + 1U >= argv.size()) {
                error = std::string(what);
                return std::nullopt;
            }
            return std::string_view(argv[++i]);
        };

        if (arg == "--input") {
            const auto value = next("--input requires a path");
            if (!value) {
                return std::nullopt;
            }
            args.input = std::filesystem::path(*value);
            have_input = true;
        } else if (arg == "--output") {
            const auto value = next("--output requires a path");
            if (!value) {
                return std::nullopt;
            }
            args.output = std::filesystem::path(*value);
            have_output = true;
        } else if (arg == "--seed") {
            const auto value = next("--seed requires a number");
            if (!value) {
                return std::nullopt;
            }
            if (!value->empty() && value->front() == '-') {
                error = "--seed must not be negative: " + std::string(*value);
                return std::nullopt;
            }
            std::uint64_t seed = 0;
            const auto* begin = value->data();
            const auto* end = begin + value->size();
            const auto [ptr, ec] = std::from_chars(begin, end, seed);
            if (ec != std::errc{} || ptr != end) {
                error = "--seed is not a number: " + std::string(*value);
                return std::nullopt;
            }
            args.seed = seed;
        } else if (arg == "--help" || arg == "-h") {
            error.clear();
            return std::nullopt;
        } else {
            error = "unknown argument: " + std::string(arg);
            return std::nullopt;
        }
    }

    if (!have_input) {
        error = "--input is required";
        return std::nullopt;
    }
    if (!have_output) {
        error = "--output is required";
        return std::nullopt;
    }
    return args;
}

}  // namespace

int main(int argc, char** argv) {
    const std::span<const char* const> arguments(const_cast<const char* const*>(argv) + 1,
                                                 static_cast<std::size_t>(argc > 0 ? argc - 1 : 0));

    std::string error;
    const auto args = parse_args(arguments, error);
    if (!args) {
        return usage(error);
    }

    auto events = ems::core::read_journal(args->input);
    if (!events) {
        // No stack unwind noise: a malformed input journal is a data problem,
        // and the line number in the message is what actually helps.
        std::cerr << "ems-slice: " << events.error().to_string() << '\n';
        return kExitInputError;
    }

    ems::core::DeterministicIds ids{args->seed};
    const auto output = ems::it::run_slice(events.value(), ids);

    if (auto status = ems::core::write_journal(args->output, output); !status) {
        std::cerr << "ems-slice: " << status.error().to_string() << '\n';
        return kExitInputError;
    }
    return kExitOk;
}

File: java/ems-it/src/main/java/io/crossasset/ems/it/slice/SliceMain.java
/*
 * Copyright (c) 2026 Cross-Asset EMS Contributors.
 * Licensed under the Apache License, Version 2.0.
 */
package io.crossasset.ems.it.slice;

import io.crossasset.ems.core.journal.DeterministicIds;
import io.crossasset.ems.core.journal.JournalEvent;
import io.crossasset.ems.core.journal.MalformedJournalException;
import io.crossasset.ems.transport.journal.JournalTransport;
import io.crossasset.ems.transport.journal.Transport;
import java.io.PrintStream;
import java.io.UncheckedIOException;
import java.nio.file.Path;
import java.util.List;
import org.jspecify.annotations.Nullable;

/**
 * {@code ems-slice --input <journal> --output <journal> [--seed <n>]}
 *
 * <p>A pure function from an input journal to an output journal. No network, no clock, no
 * filesystem beyond those two paths. The Rust and C++ binaries accept the same arguments and must
 * produce byte-identical output — see {@code conformance/README.md}.
 *
 * <p>Exit codes: {@code 0} success, {@code 1} the input journal could not be read, {@code 2} a
 * usage error. These are part of the contract too: the harness distinguishes "this implementation
 * rejected the input" from "this implementation was invoked wrongly".
 */
public final class SliceMain {

  private static final int EXIT_OK = 0;
  private static final int EXIT_INPUT_ERROR = 1;
  private static final int EXIT_USAGE = 2;

  private SliceMain() {}

  public static void main(String[] args) {
    System.exit(run(args, System.err));
  }

  /**
   * Testable entry point. Returns the exit code instead of calling {@link System#exit}, so the
   * tests can assert on it.
   */
  public static int run(String[] args, PrintStream err) {
    @Nullable Path input = null;
    @Nullable Path output = null;
    long seed = 0;

    for (int i = 0; i < args.length; i++) {
      String arg = args[i];
      switch (arg) {
        case "--input" -> {
          if (++i >= args.length) {
            return usage(err, "--input requires a path");
          }
          input = Path.of(args[i]);
        }
        case "--output" -> {
          if (++i >= args.length) {
            return usage(err, "--output requires a path");
          }
          output = Path.of(args[i]);
        }
        case "--seed" -> {
          if (++i >= args.length) {
            return usage(err, "--seed requires a number");
          }
          try {
            seed = Long.parseLong(args[i]);
          } catch (NumberFormatException e) {
            return usage(err, "--seed is not a number: " + args[i]);
          }
          if (seed < 0) {
            return usage(err, "--seed must not be negative: " + seed);
          }
        }
        case "--help", "-h" -> {
          return usage(err, null);
        }
        default -> {
          return usage(err, "unknown argument: " + arg);
        }
      }
    }

    if (input == null) {
      return usage(err, "--input is required");
    }
    if (output == null) {
      return usage(err, "--output is required");
    }

    // The slice talks to a Transport, not to files. JournalTransport is the
    // deterministic implementation the conformance gate runs; an Aeron-backed
    // one is the live path, and nothing here can tell them apart.
    // See docs/decisions/0006-abstract-transport-journal-first.md.
    try (Transport transport = new JournalTransport(input, output)) {
      List<JournalEvent> events = transport.drain();
      for (JournalEvent event : new SliceRunner(new DeterministicIds(seed)).run(events)) {
        transport.publish(event);
      }
      transport.flush();
      return EXIT_OK;
    } catch (MalformedJournalException e) {
      // No stack trace: a malformed input journal is a data problem, and a
      // stack trace here would bury the line number that actually helps.
      err.println("ems-slice: malformed input journal: " + e.getMessage());
      return EXIT_INPUT_ERROR;
    } catch (UncheckedIOException e) {
      err.println("ems-slice: " + e.getMessage());
      return EXIT_INPUT_ERROR;
    }
  }

  private static int usage(PrintStream err, @Nullable String message) {
    if (message != null) {
      err.println("ems-slice: " + message);
    }
    err.println(
        """
        usage: ems-slice --input <journal> --output <journal> [--seed <n>]

          --input   input event journal (JSONL)
          --output  output event journal (JSONL), overwritten
          --seed    identifier generator seed, default 0

        A pure function from input journal to output journal. See conformance/README.md.""");
    return EXIT_USAGE;
  }
}
